File: C++/RayTracer/scene.cpp
#include "scene.h"
#include "frame_buffer.h"
#include "utilMethods.h"
#include "intersection.h"
#include <iostream>

using namespace std;

// Load a Layout (.rtl) file
void Scene::Load(char* file)
{
	FILE* pSceneFile = fopen(file, "r");
	
	if (!pSceneFile) {
		cout << "Failed to load " << file << "." << endl;
		char* errorMsg = strerror(errno); // GNU-specific version, Linux default
		printf("Error %s", errorMsg);
	}
	else
		cout << "Successfully loaded " << file << "." << endl;

	// Get sphere, triangle mesh, and light counts
	sphereCount = 0;
	triangleMeshCount = 0;
	lightCount = 0;
	fscanf(pSceneFile, "%d %d %d\n", &lightCount, &sphereCount, &triangleMeshCount);

	sphereList = new Sphere[sphereCount];
	triangleMeshList = new TriangleMesh[triangleMeshCount];
	lightList = new Light[lightCount];

	//fseek(pSceneFile, 0L, SEEK_SET);

	// Step through the file and create/load the lights
	for (int i = 0; i < lightCount; i++)
	{
		char label;
		int type; 
		Vertex pos;
		Color col;
		
		fscanf(pSceneFile, "%c %d %f %f %f %lf %lf %lf\n", 
			&label, &type, &pos.x, &pos.y, &pos.z, &col.r, &col.g, &col.b);
		
		if (label == 'L') {
			lightList[i].isPoint = type;
			lightList[i].position = pos;
			lightList[i].color = col;
		}
	}

	// Step through the file and create/load the lights
	for (int i = 0; i < sphereCount; i++)
	{
		char label;
		Vertex pos;
		Color cAmb, cDif, cSpec;
		float r, ka, kd, ks, ns, nr, kl, kr;

		fscanf(pSceneFile, "%c %f %f %f %f %lf %lf %lf %lf %lf %lf %lf %lf %lf %f %f %f %f %f %f %f\n",
			&label, &pos.x, &pos.y, &pos.z, &r, &cAmb.r, &cAmb.g, &cAmb.b, &cDif.r, &cDif.g, &cDif.b,
			&cSpec.r, &cSpec.g, &cSpec.b, &ka, &kd, &ks, &ns, &nr, &kl, &kr);

		if (label == 'S') {
			sphereList[i].center = pos;
			sphereList[i].radius = r;
			sphereList[i].ambientColor = cAmb;
			sphereList[i].diffuseColor = cDif;
			sphereList[i].specularColor = cSpec;
			sphereList[i].ka = ka;
			sphereList[i].kd = kd;
			sphereList[i].ks = ks;
			sphereList[i].ns = ns;
			sphereList[i].nr = nr;
			sphereList[i].kReflect = kl;
			sphereList[i].kRefract = kr;
		}
	}

	// Step through the file and create/load the lights
	for (int i = 0; i < triangleMeshCount; i++)
	{
		char label;
		char meshFile[255];
		Vertex pos;
		float rotX, rotY, rotZ;
		float scale;
		Color cAmb, cDif, cSpec;
		float ka, kd, ks, ns, nr, kl, kr;

		fscanf(pSceneFile, "%c %s %f %f %f %f %f %f %f %lf %lf %lf %lf %lf %lf %lf %lf %lf %f %f %f %f %f %f %f\n",
			&label, &meshFile, &scale, &rotX, &rotY, &rotZ, &pos.x, &pos.y, &pos.z, &cAmb.r, &cAmb.g, &cAmb.b, 
			&cDif.r, &cDif.g, &cDif.b, &cSpec.r, &cSpec.g, &cSpec.b, &ka, &kd, &ks, &ns, &nr, &kl, &kr);

		if (label == 'M') {
			triangleMeshList[i].Load(meshFile, scale, rotX, rotY, rotZ, pos.x, pos.y, pos.z);
			triangleMeshList[i].ambientColor = cAmb;
			triangleMeshList[i].diffuseColor = cDif;
			triangleMeshList[i].specularColor = cSpec;
			triangleMeshList[i].ka = ka;
			triangleMeshList[i].kd = kd;
			triangleMeshList[i].ks = ks;
			triangleMeshList[i].ns = ns;
			triangleMeshList[i].nr = nr;
			triangleMeshList[i].kReflect = kl;
			triangleMeshList[i].kRefract = kr;
		}
	}

	cout << "Number of lights Loaded: " << lightCount << endl;
	cout << "Number of spheres Loaded: " << sphereCount << endl;
	cout << "Number of triangle meshes Loaded: " << triangleMeshCount << endl;
}

void Scene::RayTrace(FrameBuffer* fb) {
	//Dimensions of image plane
	const int numPixelsX = fb->GetWidth();
	const int numPixelsY = fb->GetHeight();
	int prevPercentDone = 0;

	//Iterate over all virtual pixels in frame buffer (image plane)
	//Shift image plane to have center at z = 0 in world space
	for (int i = 0; i < numPixelsX ; i++) {
		for (int j = 0; j < numPixelsY; j++) {
			//Map framebuffer pixels to image plane coordinates
			float x = ((imagePlaneWidth / (float)numPixelsX) * (float)i) - (imagePlaneWidth / 2.0f);
			float y = ((imagePlaneHeight / (float)numPixelsY) * (float)j)- (imagePlaneHeight / 2.0f);

			//Compute center of virtual pixel
			float pixelWidth = (float)imagePlaneWidth / (float)numPixelsX;
			float pixelHeight = (float)imagePlaneHeight / (float)numPixelsY;
			float xCenter = x + pixelWidth / 2.0f;
			float yCenter = y + pixelHeight / 2.0f;

			//printf("pixel (%d, %d), coord (%f, %f) center @ (%f, %f)\n", i, j, x, y, xCenter, yCenter);

			//Get vector u = ray cast from center of projection through virtual pixel center
			Vertex pPix = Vertex();
			pPix.x = xCenter; 
			pPix.y = yCenter; 
			pPix.z = imagePlaneZ;

			//Point of projection is 0, 0, 0 for this assignment
			Vertex pPrp = Vertex();
			pPrp.x = 0.0; 
			pPrp.y = 0.0; 
			pPrp.z = 0.0;

			Vector u = subtract(pPix, pPrp);
			u.normalize();

			Color pixelColor = ShootRay(pPrp, u, maxDepth, attenuationFactor);

			fb->SetPixel(i, j, pixelColor);

			int percentDone = (int)(100 * ((float)(i * numPixelsX + j) / (float)(numPixelsX * numPixelsY)));
			
			if (percentDone % 1 == 0 && prevPercentDone != percentDone)
				printf("%d ", percentDone);
			prevPercentDone = percentDone;
		}
	}
}

Color Scene::ShootRay(Vertex P0, Vector u, int depth, float attenuationFactor) {
	
	Color result = Color();

	//intersection closest to camera
	Intersection minIntersection = Intersection();
	minIntersection.distance = FLT_MAX;

	//Find any intersections with spheres
	for (int i = 0; i < sphereCount; i++) {
		Sphere sphere = sphereList[i];
		Intersection intersection = intersectSphere(P0, u, sphere);
		if (intersection.distance < minIntersection.distance && intersection.distance > FLT_MIN + 1) {	
			minIntersection = intersection;
		}
	}

	//Find any intersections with triangle meshes
	for (int i = 0; i < triangleMeshCount; i++) {
		TriangleMesh tMesh = triangleMeshList[i];
		Intersection intersection = intersectTriangleMesh(P0, u, tMesh);
		if (intersection.distance < minIntersection.distance && intersection.distance > FLT_MIN + 1) {
			minIntersection = intersection;
		}
	}

	//Base case 1 => No intersection
	if (minIntersection.distance == FLT_MAX) {
		return result;
	}

	//Flip normal if inside
	Vector N = minIntersection.normal;
	if (minIntersection.isInside) {
		N = multiply(N, -1.0);
	}
	//If intersection, continue computations and ray casting on that intersection
	Color localIntensity = Color();

	depth = depth - 1;
	if (depth > 0) {
		//if(depth < 3)
			//printf("%d\n", depth);
		//Base case => If non reflective or refractive surface
		if (minIntersection.obj.kReflect == 0 && minIntersection.obj.kRefract == 0) {
			return phongIllumination(minIntersection.point, N, minIntersection.obj, lightList, lightCount, Ia);
		}
		//If reflective surface
		if (minIntersection.obj.kReflect > 0) {
			if (!minIntersection.isInside) {
				localIntensity = phongIllumination(minIntersection.point, N, minIntersection.obj, lightList, lightCount, Ia);
				result = localIntensity;

				//Shoot reflected ray
				//Calculate reflection vector
				Vector N2 = multiply(N, dot(multiply(u, 2), N));
				Vector R = multiply(subtract(u, N2), 1.0);
				R.normalize();

				//Reduce attenuation factor
				float krg = attenuationFactor;

				//Shoot reflected ray from intersection point in direction of reflected ray
				Color I = ShootRay(minIntersection.point, R, depth, krg);

				//If there was no intersection I is black (0,0,0) so we need no condition
				result = result + I * minIntersection.obj.kReflect * (double)krg;
			}
		}

		//Shoot refracted ray
		//If object is refractive
		if (minIntersection.obj.kRefract > 0) {
			//Compute refraction ray
			//Compute cos of angle between u and surface normal = thetai
			float UdotN = dot(u, N);
			float cosThetai = UdotN / (u.magnitude() * N.magnitude());
			float cosThetai2 = 1.0 - pow(cosThetai, 2);
			//Compute cos of angle between refracted ray and (negative) surface normal = thetaR
			float radicand = (1.0 - (cosThetai2));
			float cosThetaR = sqrt(radicand);

			Vector v1 = multiply(u, minIntersection.obj.nr);
			float s1 = cosThetaR - (minIntersection.obj.nr * cosThetai);
			Vector v2 = multiply(N, s1);
			
			//Refraction ray
			Vector T = subtract(v1, v2);
			T.normalize();

			minIntersection.point = add(minIntersection.point.toVector(), multiply(T, 0.5)).toVertex();

			//Reduce attenuation factor
			float krg = attenuationFactor ;

			//Shoot refracted ray from intersection point in direction of refrefracted lected ray
			minIntersection.isInside = !minIntersection.isInside;
			Color I = ShootRay(minIntersection.point, T, depth, krg);

			//If there was no intersection I is black (0,0,0) so we need no condition
			result = result + I * minIntersection.obj.kRefract * (double)krg;
		}
		return result;
	}
	return result;
}

File: Go/RayTracer/src/scene/scene.go
package scene

import (
	"log"
	"math"
	"prims"
	"util"
	"sync"
)

//Represents scene containing objects and lights
type Scene struct {
	Lights   	[]prims.Light	//List of lights in the scene
	Spheres     []prims.Sphere	//List of objects in the scene
	Cam			*util.Vertex	//Location of camera in scene
	Depth    	int				//Depth of trace for ray tracing
	IPz      	float64			//Z location of image plane
	IPWidth  	int				//Width of image plane
	IPHeight 	int				//Height of image plane
	NumPixX		int				//Number of image plane pixels in X
	NumPixY		int				//Number of image plane pixels in Y
	Ambient  	*prims.Color	//Ambient lighting color in the scene
}


//Ray trace the scene with the given scene parameters
func (s *Scene) RayTrace(numPixX int, numPixY int) [][]prims.Color {
	if numPixX % s.IPWidth != 0 || numPixY % s.IPHeight != 0 {
		log.Fatal("image resolution must be a multiple of image plane dimensions")
	}

	pixColorData := make([][]prims.Color, numPixX)
	for i := range pixColorData {
		pixColorData[i] = make([]prims.Color, numPixY)
	}
	
	scd := &prims.SafeColorData { 
		Data : pixColorData, 
	}

	traceWithParallelization(numPixX, numPixY, s, scd)
	return scd.Data
}

func traceWithoutParallelization(numPixX, numPixY int, s *Scene, scd *prims.SafeColorData){
	//Ray trace for each image plane pixel
	for i := 0; i < numPixX; i++ {
		for j := 0; j < numPixY; j++ {
			r := computeRayForTrace(s, numPixX, numPixY, i, j)
			c := trace(r, 5, 1, s)
			scd.Set(i, j, c)
		}
	}
}

func traceWithParallelization(numPixX, numPixY int, s *Scene, scd *prims.SafeColorData){
	var wg sync.WaitGroup 
	//Ray trace for each image plane pixel
	for i := 0; i < numPixX; i++ {
		ii := i 

		//Start goroutines to each fill a row of color data
		wg.Add(1)
		go func(){
			defer wg.Done()
			for j := 0; j < numPixY; j++ {
				r := computeRayForTrace(s, numPixX, numPixY, ii, j)
				c := trace(r, 5, 1, s)
				scd.Set(ii, j, c)
			}
		}()
	}
	wg.Wait()
}

//Compute the ray location and direction based on input pixel
func computeRayForTrace(s *Scene, numPixX, numPixY, i, j int) *prims.Ray {
	//Map number of image plane pixels to locations on image plane
	x := ((float64(s.IPWidth) / float64(numPixX)) * float64(i)) - (float64(s.IPWidth) / 2.0)
	y := ((float64(s.IPHeight) / float64(numPixY)) * float64(j)) - (float64(s.IPHeight) / 2.0)

	//Get center of IP pixel
	pixW := float64(s.IPWidth) / float64(numPixX)
	pixH := float64(s.IPHeight) / float64(numPixY)
	xCenter := x + pixW / 2.0
	yCenter := y + pixH / 2.0

	//Compute ray direction vector = IP pixel pos - point of projection (camera pos)
	pPix := util.Vertex{ xCenter, yCenter, s.IPz, 1 }
	u := util.VertSub(&pPix, s.Cam)
	u.Normalize()

	r := &prims.Ray{
		P0: s.Cam,
		U:  u,
	}
	
	return r
}

//Shoot a ray, compute reflections and refractions and total illumination
func trace(r *prims.Ray, depth int, atten float64, scene *Scene) prims.Color {
	finalColor := prims.Color{}

	//Get closest intersection
	minI := findMinIntersection(r, scene)

	if minI.T == math.MaxFloat64 {
		return finalColor
	}

	//If ray inside object, flip surface normal
	if minI.IsInside {
		minI.N.Scale(-1)
	}

	//Decrement depth of trace
	depth--
	if depth >= 0 {
		//Get illumination at intersection point
		localI := prims.Phong(&minI, scene.Lights, scene.Cam, *scene.Ambient)

		//if intersected object isn't reflective or refractive (ray absorbed)
		if minI.Sphere.Props.KR == 0 && minI.Sphere.Props.KT == 0 {
			return localI
		}

		//if reflective surface
		if minI.Sphere.Props.KR > 0 {
			//Shoot reflected ray from intersection in dir of reflection
			reflected := computeReflectionRay(*r.U, &minI)
			I := trace(reflected, depth, atten, scene)

			finalColor = prims.ColorAdd(localI, finalColor, I)
			finalColor.Scale(minI.Sphere.Props.KR * atten)
		}

		//If object is refractive
		if minI.Sphere.Props.KT > 0 {
			transmitted := computeTransmittedRay(*r.U, &minI)
			I := trace(transmitted, depth, atten, scene)

			finalColor = prims.ColorAdd(finalColor, I)
			finalColor.Scale(minI.Sphere.Props.KT * atten)
		}
		return finalColor
	}
	return finalColor
}

//Find the closest point of intersection along the ray
func findMinIntersection(r *prims.Ray, scene *Scene) prims.Intersection {
	//Find closest intersection
	minI := prims.Intersection{}
	minI.T = math.MaxFloat64

	for _, s := range scene.Spheres {
		i := s.Intersect(r)	
		if i.T < minI.T && i.T > math.SmallestNonzeroFloat64 + 0.0000001 {
			minI = i	
		}
	}

	//Flip normal if vector was cast into object
	if minI.IsInside {
		minI.N.Scale(-1)
	}
	
	return minI
}

//Compute the transmitted (refracted) ray direction and point of origin
func computeTransmittedRay(u util.Vector, minI *prims.Intersection) *prims.Ray {
	//Compute cos of angle between ray direction and surface normal = thetaI
	u.Normalize()
	UDotN := util.Dot(&u, minI.N)
	cosThetaI := UDotN / (u.Magnitude() * minI.N.Magnitude())
	cti2 := 1 - math.Pow(cosThetaI, 2)
	//Compute cos of angle between refracted ray and (negative) surface normal = thetaR
	radicand := (1 - cti2)
	cosThetaR := math.Sqrt(radicand)

	v1 := util.MultiplyByScalar(u, minI.Sphere.Props.Nr)
	s1 := cosThetaR - (minI.Sphere.Props.Nr * cosThetaI)
	v2 := util.MultiplyByScalar(*minI.N, s1)

	//Refraction (transmitted) ray
	T := util.VecSub(&v1, &v2)
	T.Normalize()

	//Shift intersection point to in direction of transmitted ray 
	//to avoid immediate collision with object
	shifted := util.MultiplyByScalar(*T, 0.5)
	minI.P = util.VertAdd(minI.P, shifted.ToVertex())
	minI.IsInside = !minI.IsInside;

	transmitted := &prims.Ray {
		P0: minI.P,
		U:  T,
	}

	return transmitted
}

//Compute the reflected ray direction and point of origin
func computeReflectionRay(u util.Vector, minI *prims.Intersection) *prims.Ray {
	//Compute direction of reflection vector
	//R = U - 2N * (U dot N)
	UdotN := util.Dot(&u, minI.N)
	TwoN := util.MultiplyByScalar(*minI.N, 2)
	TwoN.Scale(UdotN)
	R := util.VecSub(&u, &TwoN)
	R.Normalize()

	//Shift intersection point in direction of reflection to avoid "surface acne"
	shifted := util.MultiplyByScalar(*R, 0.5)
	minI.P = util.VertAdd(minI.P, shifted.ToVertex())

	reflected := &prims.Ray {
		P0: minI.P,
		U:	R,
	}

	return reflected
}